### ISSUE TITLE
Allowed format/characters for member names

### ISSUE THREAD

**Issue by cziegenberg** [json-api/json-api#567]
I thought about the format of member names and the following questions came into my mind:
- Is there a format for (top-level) member names? 
- Which characters are allowed (at special positions)? 
- Are member names case sensitive? 
- Can I use integers as members?

Suggestions:
- Periods (U+002E FULL-STOP, ".") aren't allowed, because it's used as the relationship path separator.
- Commas (U+002C COMMA, ",") aren't allowed, because it's used in the include parameter as a separator for multiple relationship paths.
- Members should not start with a digit ("0-9"), as this could be a problem when mapping API to the code, e.g. in PHP and C# object members of this type aren't allowed.
- Members should not contain a minus (U+002D HYPHEN MINUS, "-"), as this could be a problem when mapping API to the code, e.g. in PHP and C# object members of this type aren't allowed.
- Members should not contain a colon (U+003A COLON, ":"), as this could be a problem when mapping API to the code, e.g. in PHP and C# object members of this type aren't allowed. Also they could be reserved for later use in the include parameter?!
- When checking for duplicate members, members should be handled case insensitive, because they name a resource, which has special semantics, and this is independent from the spelling. When using member names in a request (e.g. with the include parameter), they should be handled case sensitive, because this would allow duplicate content for different spellings.
- Theoretically it isn't important if lower- or upper-case characters are used for the member names, but I'd suggest to always use lower-case, as this is the current format of the standard members and it would simplify the use by API clients.
- Underscores (U+005F LOW LINE, "_") should be allowed, as I don't know problems with underscores in member variables of any language. They could be used as the first character, but IMHO that wouldn't be good for readability.

Resulting format:
^[a-z]+[a-z0-9_]*$

**Comment by passcod** [comment]
Are you talking about what is covered here? http://jsonapi.org/recommendations/#naming Also probably see #571.

**Comment by dgeb** [comment]
I think it's probably a good idea for us to restrict use of periods and commas in names, since both could lead to ambiguity in interpreting certain query params.

I'm also tempted to not allow hyphens at the beginning of names, and go back to the assumption that default sort criteria is ascending (instead of requiring a plus prefix). This has proven somewhat annoying to implement due to URL encoding.

However, there's no reason to limit the names used in a serialization format based on language requirements. Every language has its own (unique) requirements, and every implementation will need a serialization / deserialization layer.

Marking for consideration by 1.0.

**Comment by cziegenberg** [comment]
@passcod Yes, that's what I found later - but that's only a recommendation, although some characters (like periods, see list above) are problematic...

**Comment by cziegenberg** [comment]
@dgeb Oh, that right. I missed the -/+ characters for ordering in the list of problems. I agree, that every language has different requirements, but MANY languages have the mentioned problems and a restriction would simplify implementations.

**Comment by passcod** [comment]
Also maybe consider that member names could contain Unicode beyond the U+00FF (ASCII) range. It's probably not great API design, but it could happen.

**Comment by cziegenberg** [comment]
I also used an extended alphabet and would never use anything but ASCII characters for members, but I agree that this should be possible. Why not use emoji as members? 🐙

I adjusted my suggestion, to not conflict that much with the current recommendation - some things are the decision of the API designer... So I concentrated on UNICODE support and future versions/features.

Updated suggestion:
- Members MUST NOT contain periods (U+002E FULL-STOP, "."), because this character is used as the relationship path separator.
- Members MUST NOT contain commas (U+002C COMMA, ",") , because this character is used in the include parameter as a separator for multiple relationship paths.
- Members MUST NOT start with a minus (U+002D HYPHEN MINUS, "-") or a plus (U+002B PLUS, "+"), as they have a special meaning for ordering. Also the underscore (U+005F LOW LINE, "_") is not allowed as the first character (reserved for later use).
- Lower- or upper-case characters can be used for members. When processing members (e.g. in the include parameter), the server MUST handle them case-sensitive. When checking for duplicate members, they MUST be handled case insensitive, because using members with the "same" name but different spellings would be a really bad practice...
- Members MUST NOT contain other ASCII characters than "0-9", "a-z", "A-Z", minus (U+002D HYPHEN MINUS, "-") and underscores (U+005F LOW LINE, "_"). Other ASCII characters are not allowed (control characters), have a special meaning (see above) or are reserved for later use.
- Members must have a minimum length of one character.
- Members can contain UNICODE, but using ASCII characters is recommended, as this could be a problem when mapping the API to the code on server and/or client side UNICODE characters can also cause problems, because not all servers can handle UNICODE parameters in URLs correctly (had this problem with some IIS versions).
- Members SHOULD NOT start with a digit ("0-9"), as this could be a problem when mapping the API to the code on server and/or client side.
- Members can use minus (U+002D HYPHEN MINUS, "-") or underscores (U+005F LOW LINE, "_") as a separator. A minus could cause problems when mapping the API to the code on server and/or client side, while an underscore isn't a problem in most languages. On the other hand the minus is better readable than the underscore.

Resulting format:
^([a-zA-Z0-9]|[^[:ascii:]])+([-a-zA-Z0-9_]|[^[:ascii:]])*$

**Comment by ethanresnick** [comment]
I'm with @ziege and @dgeb re not allowing commas, periods, or a leading +/- in member names. I also support @ziege's suggestions to:
- add a _recommendation_ to only have ascii characters in property names, which makes sense for our character encoding–illiterate world
- require that property names can't be the empty string, which would be madness in query parameters

I'm not in favor of:
- reserving leading underscores, since we've rejected prefixing reserved keys repeatedly in the past.
- allowing case-insensitivity in some (and yet not all) contexts. Most languages are case sensitive, so anything but full case sensitivity in the spec seems bound to introduce implementation bugs.
- restricting ascii control characters seems weird
- preventing all ascii control characters, which would explicitly prevent interoperability with json-ld

**Comment by passcod** [comment]
I'd be wary of introducing _anything_ which uses programming languages, software, and tools as a justification. E.g. introducing a constraint because of the behaviour of _some_ servers (IIS) lies not with the spec, but with the developer implementing such an API for such a server.

To add to @ethanresnick's list, it would probably be wise to further forbid property names that are **all whitespace**, for the same reason as stated.

What is the opinion / recommendation of the spec (and/or its people) for whitespace _within_ property names?

**Comment by cziegenberg** [comment]
@ethanresnick:

I not only disallowed plus, minus and the underscore as the first character, but all ASCII characters that are not a-z or A-Z (or 0-9, although not recommended) - UNICODE characters would be allowed without any restriction (although not recommended). The reason for this is that these characters already have a special meaning (plus and minus for ordering, period and comma as separators) or could be used for later features of the API. Reserved keys are defined by the spec, so a exception of this rule could be defined for them (although I don't think this a good idea and no more necessary, if the proposal #573 and #574 are realized).

I perhaps didn't expressed it clearly enough, but I suggested full case-sensitivity for the members. All I added to this is that there must not be members with the same name but different spellings, e.g. "id", "Id", "iD", and "ID". This could result in implementation problems, e.g. when mapping these member in a language that doesn't isn't case-sensitive. So I think we both want the same. :)

Why don't you want to restrict control characters? Do you really want to allow line feeds, tabs or null in members? I don't think that this makes sense, but one could see it as a weird decision of the API designer... So one could turn it into a recommendation.

I didn't know JSON-LD, but after a fast check they seem to allow everything but an "@" as the first character - so it does not only allow control characters, but also members starting with minus, plus, period and comma... So interoperability it a general problem here.

@passcod:

You're right. A buggy web server should not limit this spec, but that's only an example which can be removed of course (if the API uses UNICODE members, the server has to support it). The main reason to prefer ASCII is the client side, because it could be difficult for some clients to use members in Hebrew or Chinese...

Members with ASCII space characters are disallowed by my definition, but there are a lot of UNICODE white space characters (even invisible ones, without any space) that would be allowed. I wouldn't restrict these UNICODE characters and also see it "as a weird decision of the API designer".

**Comment by passcod** [comment]
> This could result in implementation problems, e.g. when mapping these member in a language that doesn't isn't case-sensitive.

Again, lets not concern ourselves too much with what _might_ be an implementation problem in _some_ languages. In fact, consider what happens for languages where case differences _are_ significative… like ruby. The best move, I'd say, is not to play.

**Comment by cziegenberg** [comment]
Okay, I generally agree, that the spec should not be that restrictive and that problems that can occur should only be listed as recommendation, not as a must - but in any case I would reserve most special ASCII characters for later versions of the spec (to not break implementations later).

**Allowed ASCII characters:**
- a-z
- A-Z
- 0-9
- U+0020 SPACE, " " (except as first character)
- U+002D HYPHEN-MINUS, "-" (except as first character)
- U+005F LOW LINE, "_" (except as first character)
- U+005E CIRCUMFLEX ACCENT, "^" (except as first character)
- U+0060 GRAVE ACCENT, "`" (except as first character)

**Reserved ASCII characters (currently in use):**
- U+002B PLUS SIGN, "+" (used for ordering)
- U+002C COMMA, "," (used separator for multiple relationship paths)
- U+002E PERIOD, "." (used as relationship path separators)
- U+005B LEFT SQUARE BRACKET, "[" (use in sparse fieldsets)
- U+005D RIGHT SQUARE BRACKET, "]" (used in sparse fieldsets)

**Reserved ASCII characters (punctation, currecy and math symbols) for later use:**
- U+0021 EXCLAMATION MARK, "!"  
- U+0022 QUOTATION MARK, '"'
- U+0023 NUMBER SIGN, "#"
- U+0024 DOLLAR SIGN, "$"
- U+0025 PERCENT SIGN, "%"
- U+0026 AMPERSAND, "&"
- U+0027 APOSTROPHE, "'"
- U+0028 LEFT PARENTHESIS, "("
- U+0029 RIGHT PARENTHESIS, ")"
- U+002A ASTERISK, "*"
- U+002F SOLIDUS, "/"
- U+003A COLON, ":"
- U+003B SEMICOLON, ";"
- U+003C LESS-THAN SIGN, "<"
- U+003D EQUALS SIGN, "="
- U+003E GREATER-THAN SIGN, ">"
- U+003F QUESTION MARK, "?"
- U+0040 COMMERCIAL AT, "@"
- U+005C REVERSE SOLIDUS, "\"
- U+007B LEFT CURLY BRACKET, "{"
- U+007C VERTICAL LINE, "|"
- U+007D RIGHT CURLY BRACKET, "}"
- U+007E TILDE, "~"

**This results in the following rules:**
- Members MUST start and end with one of the following ASCII characters: a-z, A-Z, 0-9, or any NON-ASCII character.
- Members MUST contain only non-reserved ASCII characters or NON-ASCII characters
- Members MUST be handled case-sensitive.
- Members MUST contain at least one character.

**And some recommendations, to avoid potential client/server problems:**
- Members SHOULD NOT start with a digit ("0-9").
- Members SHOULD NOT contain NON-ASCII characters.
- Members SHOULD NOT contain ASCII control characters.
- Members SHOULD NOT use the "same" name in different spellings (e.g. "id" and "ID").

**Comment by cziegenberg** [comment]
If you need an example for the later use of special characters, let's imagine that the next version of the spec defines the filter parameter in detail as follows...

**Example format:**
filter=(book.author.name = "J. R. R. Tolkien" && (book.pages >= 300 || book.title ~= "Hobbit"))

**Used characters:**
Parenthesis:
- (
- )

Quotes:
- "
- '

Logical operators:
- &&
- ||

Comparison operators:
- "=", equal
- ">", greater
- ">=", greater or equal
- "<", less than
- "<=", less than or equal
- "~=", contains
- "|=", starts with
- "=$", ends with

Please note: This is no proposal, just an example! :)

**Comment by hhware** [comment]
The only thing which I would change in the last version of this proposal is case-sensitivity for member names. I think it would be good to make them case-insensitive for all purposes. Case-sensitivity would only bring pain (and tech support burden in commercial setting) in case of APIs, when the server and multitude of clients are not controlled by the same people and when teams of various levels of expertise and attention to detail have to interoperate.

**Comment by passcod** [comment]
Again, I think that as there is both a case for and against case-insensitivity, this spec should specify case-sensitivity (or not specify anything as case-sensitivity is the "default" in JSON/strings), and then _your implementation_ should specify otherwise. You do write documentation on how your API works, including the additional constraints to the base spec you enforce, right?

**Comment by cziegenberg** [comment]
I also prefer all lower-case members, to avoid possible problems in clients - that's why I added this to the recommendations - but I have been convinced, that some people prefer to also use upper-case characters (on their own risk).

I also thought about UNICODE support and my own language (German), where not all characters (the "ß") have a lower- and upper-case version. I'm sure other languages/scripts also have such exceptions, so a must for lower-case here would perhaps impossible to fulfill when UNICODE is used.

I think everything is covered by the recommendations and if an API wants to support upper-/mixed-case members and/or UNICODE, then it's clearly against the recommendations and one has to live with possible client problems.

**Comment by jonkoops** [comment]
I believe we should also agree upon a naming convention for multi-word identifiers, currently the specification recommends the use of a hyphen to separate words:

> ## Recommendations for Naming
> 
> It is recommended that resource types, attribute names, and association names be "dasherized"; i.e. consist of only lower case alphanumeric characters and dashes (U+002D HYPHEN-MINUS, "-").
> 
> It is also recommended that resource types be pluralized. Dasherized and pluralized resource types can be used as URL segments without translation, as discussed below.

Source: http://jsonapi.org/recommendations/#naming

I think it would be a good idea to make this part of the specification instead of just a recommendation. I've already noticed this [causing some confusion](https://github.com/eneuhauser/ember-json-api/issues/8) amongst implementers.

My fear is that we might end up with a lot of different syntax in APIs that clients need to be able to consume, which might complicate the client code. If there are no significant detrimental effects I think it would be wise to make this recommendation a MUST or at least a highly enforced SHOULD.

**Comment by passcod** [comment]
Probably a SHOULD, to allow implementations to be overridden in case of need. Or maybe specify that the default behaviour be dasherising, and specify that clients should allow that behaviour to be overridden.

**Comment by jonkoops** [comment]
@passcod Yes, I was thinking the same but I honestly cannot think of a case where someone would legitimately **need** to overrule this convention.

**Comment by passcod** [comment]
Consistency with other APIs/code/established style guides within a company/team/etc, for example.

**Comment by cziegenberg** [comment]
Or because the minus is problematic in some languages, e.g. when mapping to variables.

For example in PHP you can directly read JSON into an object structure, but those member names with minus are not allowed there.

**Comment by jonkoops** [comment]
> Consistency with other APIs/code/established style guides within a company/team/etc, for example.

I don't think this is the case with JSON API. When people implement it shouldn't deal with direct raw data from the server, but with a library that exposes an API to interact with it. This library should also take care of transforming the format of the entity to one native to the language they're using.

> For example in PHP you can directly read JSON into an object structure, but those member names with minus are not allowed there.

Seems to me that this could be easily resolved by running it trough a function that transforms the names to a format that PHP can understand. If I am correct you are referring to the `json_decode` function? It should parse the response even if member names are dashterized (see http://3v4l.org/GIrmC).

**Comment by cziegenberg** [comment]
@jonkoops: I just wanted to show you, that there can be problems with some languages/implementations. In case of PHP it normally doesn't work - you can decode the JSON string but not access the members, and you cannot use the same member names in your own classes. But I don't want to discuss that in detail here... :)

One the one hand I agree with you, because every client and server should be able to to handle all member names that are allowed by the JSON standard. It is not a problem of the spec, how this is done. So one could say that every multi-word member name needs to be separated by a "minus".

On the other hand the implementation of this spec should try to avoid known problems and to not to be too restrictive. As you can see in the first version of my suggestion, I replaced the minus with an underscore, because of possible problems with clients/servers - but I have been convinced that one shouldn't restrict the naming of members, because it's a decision of the API provider and perhaps required by the used language.

I think a naming convention is a good idea and could help if you work with multiple APIs using JSON API, but I also think that a MUST would be too restrictive. 

So I suggest to add a naming recommendation to the list, but not a requirement. And I'd still suggest to prefer the underscore over the minus to avoid known problems (but this is against the current recommendation in the spec).

**Comment by jonkoops** [comment]
I can agree a MUST is maybe a bit overboard, a strong SHOULD can be enough. However I personally think having a strict convention in place to make responses uniform is more important than possible exceptions that JSON parsers might have. Which character should be used to separate words isn't as much important as that the use of it is consistent.

As for using an underscore for naming vs a minus, I am fine with either. Although also take into account such a character should also be used for relation names in the URL i.e. `/some-resource-name` (which I think why the recommendation in the spec right now is dashterized).

**Comment by cziegenberg** [comment]
Okay, then I'll add a recommendation for it and prepare a PR for it soon (if no one else with better English knowledge wants to do it), but we still need to decide which character to recommend. Minus or underscore, both have their disadvantages...

What do you think @passcod and @hhware?

**Comment by passcod** [comment]
Hyphens are the current rec, I think both for that, for compatibility, and
for aesthetics, keep hyphens.

**Comment by hhware** [comment]
@ziege, I do no think it is that important because it is only a recommendation, it will not be enforced. Because of that, generic libraries will likely have to support all options, and specific API implementation can just be opinionated and use whatever it is preferred by the API designer.

**Comment by jonkoops** [comment]
@hhware My argument for making it required in a nutshell. If we allow developers to 'pick and choose' I am afraid we'll end up with clients with too much configuration going on or in the worst case does not support the method of separating words the API designer had in mind.

I don't think there is an API designer who wants to make this decision, nor is there a developer on client-side of things that wants to deal with that decision. I think it should be part of the specification to limit the scope of these options to alleviate trivial decision making and prevent unnecessary configuration related frustration.

If we cannot agree on a default we should at least define the scope of "all options". Since there is a [whole Wikipedia page](http://en.wikipedia.org/wiki/Naming_convention_%28programming%29#Multiple-word_identifiers) about naming conventions in programming and a client cannot possibly implement all of them.

**Comment by hhware** [comment]
> If we allow developers to 'pick and choose'

@jonkoops It has been a general position of JSON API on many topics to allow exactly that. It is hard to say whether it provides _the perfect balance_ between strictness and openness for adoption, but IMHO it cannot be claimed that this position is unreasonable or wrong. 

>  a client cannot possibly implement all of them

I agree, and the scope of this issue seems to address exactly that.

**Comment by jonkoops** [comment]
The first two paragraphs on the landing page of the spec seem to contradict that statement.

> If you've ever argued with your team about the way your JSON responses should be formatted, JSON API is your anti-bikeshedding weapon.
> 
> By following shared conventions, you can increase productivity, take advantage of generalized tooling, and focus on what matters: your application.

I agree that it is not unreasonable to want to give developers the freedom to choose. But this 'freedom' will make client implementations more complex.

To me it seems there is motivation to keep this flexible just because some people prefer to have their responses have a different cosmetic appearance, adding no real value whatsoever.

This to me is bikeshedding, making API designers make trivial decisions and now everyone using the API has to take that decision into account.

**Comment by passcod** [comment]
So, for me there's really three separate issues / considerations that have emerged here:
1. The question of whether to make these changes in the spec itself (and if so at what level, SHOULD or MUST), or make them at a recommendation level.
2. The mapping from the JSON to the language and vice versa, as implemented by the (probably generic) client.
3. The formatting of multi-word member names.

**1.** is a discussion that should probably be had _later_, after the rest is mostly decided upon. For now let's assume these are to be recommendations, and to re-discuss their inclusion in the spec proper after we actually have consensus.

I don't really have an opinion on **3.**, and for **2.** I am of two minds: either the client library deals in _strings_, with an API style like e.g. `$value = $someInstanceOfAResource->get("member-name");`, in which case there is no problem whatsoever in using absolutely anything in UTF-8 for names, or the library deals in "language tokens", with an API style like e.g. `$value = $someInstance->memberName;`, in which case it would be much easier to have a strict "lowercase with hyphens".

But I want to remark something: at this moment, there is a recommendation which reads:

> It is also recommended that resource types be pluralized. Dasherized and pluralized resource types can be used as URL segments without translation, as discussed below.

Regardless of the prescribed _format_ (here, "dasherized and pluralized"), the _spirit_ of that recommendation is that "the resource type in the JSON" and the "resource type in the URL" should be the same, as well as the names of the relationships (because those also go in the URL).

That doesn't necessarily limit the format of these members to just URL-safe chars, because they could use URL % encoding, but I think we can agree that URL encoding is pretty ugly. And if we're going to hold the "resource type" name to a particular format, say alphanums with hyphens, I'm of the mind we should make _all_ member names hold to the same format, for ease of implementation and design consistency.

So we should discuss the format that's applicable to "resource type" names, and then apply that to everything else. At the very least, that will probably mean "url safe" characters. As per [RFC 3986](http://tools.ietf.org/html/rfc3986), that will pretty much be only:

```
unreserved = ALPHA / DIGIT / "-" / "." / "_" / "~"
```

In other words: No need to bikeshed! Another RFC has already done that for us.

**Comment by cziegenberg** [comment]
@passcod To your points:

1) The listed MUSTs have to be part of the spec of course. I in my opinion the listed SHOULDs also belong to the spec, as they define a preferred standard - the only reason to not define them as MUSTs is that the spec shouldn't be too restrictive and enable e.g. the use of UNICODE if the API designer prefers it.

2) The possible mapping problems are the reason for most of the SHOULDs. Only the hyphen minus separator could be a last problem. In most languages there are ways to workaround this problem (if not, they have a general problem with JSON support), so one could ignore it.

3) To decide that, we have to consider different aspects:
- The possible implementation problems in some languages, which can be workaround but perhaps are bad for the overall usability of libraries in those languages.
- The readability of member names in the JSON API document. A hyphen minus is used in many languages for combined words (German, English...) and feels more common.
- Also the readability of URLs (when member names are mapped to URLs) is better with a hyphen minus, and a short research showed that this character is the recommended one in most cases. Also underlined links could hide the underscores.
- The existing recommendation is the hyphen minus - this is not a strong argument, but should also be considered.

In the beginning I mainly regarded the implementation side and therefor decided for the underscore as the separator, but the readability arguments for the API end user are also very important, so that I would now tend to prefer the hyphen minus, which also matches the existing recommendation. This is what I plan for the PR at the moment.

Also I'd adjust the recommendations to meet the "URL safe" characters:
- ~~Members SHOULD NOT contain NON-ASCII characters.~~
- ~~Members SHOULD NOT contain ASCII control characters.~~
- Members SHOULD consist of the following non-reserved, URL safe ASCII characters only: a-z, A-Z, 0-9, "-"

I also added the upper-case characters, although I think one could recommend to use lower-case characters only (like in the existing recommendation). The reason not to do it was the UNICODE support, but as we recommend ASCII, we can also recommend to use lower-case.

**Comment by cziegenberg** [comment]
If we recommend to only use lower-case characters, the following recommendations would replace all previous ones (because already included):
- Member names **SHOULD** start with one of the following characters: a-z.
- Member names **SHOULD** consist only of the following URL safe ASCII characters: a-z, 0-9, "-"

Two simple rules for a standardized, url safe and readable naming of members, which is just a recommendation and can be replaced with full UNICODE support (except the mentioned reserved ASCII characters), if preferred.

**Comment by passcod** [comment]
I think we could just refer to the RFC. "The unreserved character set as
defined by RFC 3986, §2.<whatever>", for example, and then quote that
definition for clarity. No need to restate it ourselves, and it creates an
explicit link to that RFC which expresses the intent.

**Comment by cziegenberg** [comment]
@passcod: I considered this in the last version of PR #601

**Comment by tkellen** [comment]
Resolved in #601
